pallas-crypto/src/nonce/mod.rs

```diff
@@ -0,0 +1,192 @@
+use crate::hash::{Hash, Hasher};
+
+/// A nonce generator function that calculates an epoch nonce from the eta_v value (nc) of the block right before
+/// the stability window and the block hash of the first block from the previous epoch (nh).
+pub fn generate_epoch_nonce(nc: Hash<32>, nh: Hash<32>, extra_entropy: Option<&[u8]>) -> Hash<32> {
+    let mut hasher = Hasher::<256>::new();
+    hasher.input(nc.as_ref());
+    hasher.input(nh.as_ref());
+    let epoch_nonce = hasher.finalize();
+    if let Some(extra_entropy) = extra_entropy {
+        let mut hasher = Hasher::<256>::new();
+        hasher.input(epoch_nonce.as_ref());
+        hasher.input(extra_entropy);
+        hasher.finalize()
+    } else {
+        epoch_nonce
+    }
+}
+
+/// A nonce generator function that calculates a rolling nonce (eta_v) by applying each cardano block in
+/// the shelley era and beyond. These rolling nonce values are used to help calculate the epoch
+/// nonce values used in consensus for the Ouroboros protocols (tpraos, praos, cpraos).
```

Comment:
This function may panic and I think it would make sense to document this at the function's documentation level.

```suggestion
/// nonce values used in consensus for the Ouroboros protocols (tpraos, praos, cpraos).
///
/// # Panic
///
/// This function may panic if the `block_eta_vrf_0` argument is not a slice of
/// either 32 bytes or 64 bytes.
///
```